cocos2d/Platforms/Mac/CCViewMacGL.m

```diff
@@ -0,0 +1,296 @@
+    // When resizing the view, -reshape is called automatically on the main thread
+    // Add a mutex around to avoid the threads accessing the context simultaneously when resizing
+    
+    [self lockOpenGLContext];
+    
+    NSRect rect = [self convertRectToBacking:self.bounds];
+    
+    CCDirector *director = _director;
+    [director reshapeProjection: NSSizeToCGSize(rect.size) ];
+    
+    // avoid flicker
+    // Only draw if there is something to draw, otherwise it actually creates a flicker of the current glClearColor
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
+    
+    [self unlockOpenGLContext];
+}
+
+
+-(void) lockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    [glContext makeCurrentContext];
+    CGLLockContext([glContext CGLContextObj]);
+}
+
+-(void) unlockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    CGLUnlockContext([glContext CGLContextObj]);
+}
+
+// Find or make a fence that is ready to use.
+-(CCGLFence *)getReadyFence
+{
+    // First checkf oldest (first in the array) fence is ready again.
+    CCGLFence *fence = _fences.firstObject;;
+    if(fence.isReady){
+        // Remove the fence so it can be inserted at the end of the queue again.
+        [_fences removeObjectAtIndex:0];
+        return fence;
+    } else {
+        // No existing fences ready. Make a new one.
+        return [[CCGLFence alloc] init];
+    }
+}
+
+-(void)addFrameCompletionHandler:(dispatch_block_t)handler
+{
+    if(_fences == nil){
+        _fences = [NSMutableArray arrayWithObject:[[CCGLFence alloc] init]];
+    }
+    
+    CCGLFence *fence = _fences.lastObject;
+    if(!fence.isReady){
+        fence = [self getReadyFence];
+        [_fences addObject:fence];
+    }
+    
+    [fence.handlers addObject:handler];
+}
+
+-(void)beginFrame
+{
+    [self lockOpenGLContext];
+}
+
+-(void)presentFrame
+{
+    {
+        CCGLFence *fence = _fences.lastObject;
+        if(fence.isReady){
+            // If the fence is ready to be added, insert a sync point for it.
+            [fence insertFence];
+        }
+    }
+    
+    [self.openGLContext flushBuffer];
+    
+    // Check the fences for completion.
+    for(CCGLFence *fence in _fences){
+        if(fence.isCompleted){
+            for(dispatch_block_t handler in fence.handlers) handler();
+            [fence.handlers removeAllObjects];
+        } else {
+            break;
+        }
+    }
+    
+    [self unlockOpenGLContext];
+}
+
+-(GLuint)fbo
+{
+    return 0;
+}
+
+- (void) dealloc
+{
+    CCLOGINFO(@"cocos2d: deallocing %@", self);
+}
+
+#pragma mark CCGLView - Mouse Delegate
+
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```

Comment:
Yeah NSResponder could have been shorter.

Opinions on these selectors? Mash them together? Not sure if I should change something that will break user code, but if there's ever a chance for an API change, it's now. Plus I bet the the mac version doesn't get a lot of use- only a few people noticed there was no keyboard support in v3.
https://github.com/andykorth/cocos2d-iphone/blob/CCDirector/MergeSchedulerAndActionManager/cocos2d/CCResponderManager.m#L549